jbmc/src/java_bytecode/java_utils.cpp

```diff
@@ -445,3 +445,224 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+
+  const irep_idt clone_name =
+    id2string(identifier) + ".clone:()Ljava/lang/Object;";
+  code_typet::parametert this_param;
+  this_param.set_identifier(id2string(clone_name) + "::this");
+  this_param.set_base_name("this");
+  this_param.set_this();
+  this_param.type() = java_reference_type(symbol_type);
+  const code_typet clone_type({this_param}, java_lang_object_type());
+
+  parameter_symbolt this_symbol;
+  this_symbol.name = this_param.get_identifier();
+  this_symbol.base_name = this_param.get_base_name();
+  this_symbol.pretty_name = this_symbol.base_name;
+  this_symbol.type = this_param.type();
+  this_symbol.mode = ID_java;
+  symbol_table.add(this_symbol);
+
+  const irep_idt local_name = id2string(clone_name) + "::cloned_array";
+  auxiliary_symbolt local_symbol;
+  local_symbol.name = local_name;
+  local_symbol.base_name = "cloned_array";
+  local_symbol.pretty_name = local_symbol.base_name;
+  local_symbol.type = java_reference_type(symbol_type);
+  local_symbol.mode = ID_java;
+  symbol_table.add(local_symbol);
+  const auto &local_symexpr = local_symbol.symbol_expr();
+
+  code_blockt clone_body;
+
+  code_declt declare_cloned(local_symexpr);
+  clone_body.move(declare_cloned);
+
+  side_effect_exprt java_new_array(
+    ID_java_new_array, java_reference_type(symbol_type));
+  dereference_exprt old_array(this_symbol.symbol_expr(), symbol_type);
+  dereference_exprt new_array(local_symexpr, symbol_type);
+  member_exprt old_length(
+    old_array, length_component.get_name(), length_component.type());
+  java_new_array.copy_to_operands(old_length);
+  code_assignt create_blank(local_symexpr, java_new_array);
+  clone_body.move_to_operands(create_blank);
+
+  member_exprt old_data(
+    old_array, data_component.get_name(), data_component.type());
+  member_exprt new_data(
+    new_array, data_component.get_name(), data_component.type());
+
+  // Begin for-loop to replace:
+
+  const irep_idt index_name = id2string(clone_name) + "::index";
+  auxiliary_symbolt index_symbol;
+  index_symbol.name = index_name;
+  index_symbol.base_name = "index";
+  index_symbol.pretty_name = index_symbol.base_name;
+  index_symbol.type = length_component.type();
+  index_symbol.mode = ID_java;
+  symbol_table.add(index_symbol);
+  const auto &index_symexpr = index_symbol.symbol_expr();
+
+  code_declt declare_index(index_symexpr);
+  clone_body.move_to_operands(declare_index);
+
+  code_fort copy_loop;
+
+  copy_loop.init() =
+    code_assignt(index_symexpr, from_integer(0, index_symexpr.type()));
+  copy_loop.cond() = binary_relation_exprt(index_symexpr, ID_lt, old_length);
+
+  side_effect_exprt inc(ID_assign);
+  inc.operands().resize(2);
+  inc.op0() = index_symexpr;
+  inc.op1() = plus_exprt(index_symexpr, from_integer(1, index_symexpr.type()));
+  copy_loop.iter() = inc;
+
+  dereference_exprt old_cell(
+    plus_exprt(old_data, index_symexpr), old_data.type().subtype());
+  dereference_exprt new_cell(
```

Comment:
const?

Reply:
done

please note that this code was just moved, not written by me

Reply:
done, please note that I mainly moved the code and didn't write it